ISISScientificComputing/autoreduce
# Refactor WebApp variable creation

### Summary of work
#### Refactors the WebApp variable creation 

Refactors the WebApp variable creation code to re-use the new queue processor (and the real code, rather than a copy-pasted version of it!). This removes a lot of the utils. Changes touch these parts:
- Rerunning from a run summary
- Rerunning past jobs
- Configuring new jobs
- Seeing instrument variables

#### Rewrites rerun approach
Previously the rerun created ReductionRun object in the DB (via the copy pasted QP code), then submitted to the `/ReductionPending` queue. That approach was removed due to that queue not being used anymore, and to reduce the code duplication.

The new approach just submits the message to the `/DataReady` queue. 

There is a drawback here - it may not be obvious to the user whether the Re-run has been submitted, as the re-run might not show up immediately after it has been sent. This can happen if all workers are busy, and is a known limitation of the 1 queue design.

This might need an additional message in the `run_confirmation.html` or a restructuring of how the information is shown page.

#### Renames `reduction_variables` app to `instrument`
All the URLs in the ex-`reduction_variables` were in the namespace `instrument`, so it made sense that the whole thing be named instrument.

This makes for some nicer URL names such as `instrument:submit_runs`, `instrument:variables` etc.

Note the `instrument:` syntax, for when referring to a URL included from an app.

There's a few more bad URL names left in the `reduction_viewer`, but the views were too entangled to deal with.


----------
TODOs

- [x] #1098
- [x] ~~Investigate why creating experiment variables makes ALL variables, rather than just the changed ones~~
	- Left unchanged, because actually, that makes sense. When you set variables for the whole experiment it will make all variables again, some duplication but only once per experiment number, which shouldn't cause issues. We can look into it again in the future if it does
- [x] ~~"Configure New Runs" doesn't put the correct defaults when editing an experiment number/start run~~
	- This will be tackled in a follow up issue  #1133
- [x] View & selenium tests for re-run page
	- [x] Run summary (only left is sending an actual rerun)
	- [x] Rerun jobs page, part of #987
	- [x] Configure new jobs
	- [x] See instrument variables

- [x] #778
- [x] use_current_script = False path, done by removing that whole thing. The script is ignored by the reduction runner. This is captured in #1115 if we want the functionality back.
- [x] #1108

### How to test your work
The hardest part of testing this PR is setting up the env!

Before testing anything you will have to migrate the DB (`python setup.py database`) and install the basic fixtures (`python WebApp/autoreduce_webapp/manage.py loaddata super_user_fixture status_fixture two_runs` or `python setup.py fixtures`).

Make a `reduce_vars.py` for the test instrument with:
```
mkdir -p data-archive/NDXTestInstrument/user/scripts/autoreduction
echo 'standard_vars = {"variable1":"value1"}' > data-archive/NDXTestInstrument/user/scripts/autoreduction/reduce_vars.py
```

Then run ActiveMQ locally with `sudo python setup.py start` - you need to be admin for this.

Check the `credentials.ini` and make sure all is pointing locally, you can just use the values from the `test_credentials.ini` I think.

Once you've got the DB & fixtures installed you can run the queue processor & Web App, by:
- Going to `Run` tab (CTRL+SHIFT+D default keybind) in VSCode and starting the `Queue Listener` and then `Web app` - all you need to do afterwards is CTRL-click the `http://127.0.0.1:8000/` in the terminal

if not using VSCode

- Start QP with `python queue_processors/queue_processor/queue_listener.py` - this starts it without a daemon
- Start webapp with `python WebApp/autoreduce_webapp/manage.py runserver`

#### Instructions for `Rerun from Run summary`
- Navigate to the instrument runs list page
- Click on run
- Try the reset to defaults button
- Rerun it with different parameters
- See what runs come up, and with what variable values.
	- If nothing new comes in check that the QP is running, and what Queue it's been connected too.

#### Instructions for `Rerun from Rerun page`
- Similar to rerunning from run summary, but try rerunning a range of runs

#### Instructions for `Configuring new Runs` & `See instrument variables`
- You might want to use a real instrument here (e.g. MARI 27877 run is good, 27878)
	- The runs don't have to Complete, they just need to not be Skipped!
	- This means you can submit any runs
- Configure a bunch of variables for a future run range
- Then run a run inside that run range and check that they are used correctly
- Repeat for experiment number


--------

Fixes #938
Fixes #739
Fixes #852
Fixes #1098
Fixes #1112
Fixes #1119
Fixes #1108 

### Follow ups / known issues
#1133
https://github.com/ISISScientificComputing/autoreduce/issues/1154
#857
